semantic-core/src/Core/Parser.hs

```diff
@@ -55,77 +66,97 @@ identifier = choice [quote, plain] <?> "identifier" where
   plain = Token.ident coreIdents
   quote = between (string "#{") (symbol "}") (fromString <$> some (noneOf "{}"))
 
+convertSpan :: Trifecta.Span -> Source.Span
+convertSpan (Trifecta.Span s e _) = Source.Span (convertDelta s) (convertDelta e)
+  where
+    build :: Int64 -> Int64 -> Source.Pos
+    build l c = Source.Pos (fromIntegral l) (fromIntegral c)
+    convertDelta :: Trifecta.Delta -> Source.Pos
+    convertDelta = \case
+      Trifecta.Columns c _        -> build 0 c
+      Trifecta.Tab x y _          -> build 0 (Trifecta.nextTab x + y)
+      Trifecta.Lines l c _ _      -> build l c
+      Trifecta.Directed _ l c _ _ -> build l c
+
+positioned :: CoreParsing sig m t => m (t a) -> m (t a)
+positioned act = do
+  result :~ triSpan <- spanned act
+  pure (Core.annAt (convertSpan triSpan) result)
+
+
 -- * Parsers (corresponding to EBNF)
 
-core :: (TokenParsing m, Carrier sig t, Member Core sig, Monad m) => m (t Name)
+core :: CoreParsing sig m t => m (t Name)
 core = runCoreParser expr
 
-expr :: (TokenParsing m, Carrier sig t, Member Core sig, Monad m) => m (t Name)
-expr = ifthenelse <|> lambda <|> rec <|> load <|> assign
+expr :: CoreParsing sig m t => m (t Name)
+expr = positioned (ifthenelse <|> lambda <|> rec <|> load <|> assign)
```

Comment:
We don’t need a `positioned` call everywhere; annotating the expression, atom, and literal nodes should be enough.

Reply:
In case we ever revisit this: this is causing every literal to be annotated three times: once by `lit`, once by `atom`, and once by `expr`. I believe we only ever recur via `expr`, which means we could elide the latter two uses of `positioned` altogether.